sycl/doc/design/ThinLTO.md

```diff
@@ -0,0 +1,149 @@
+unit `tu_bar` that references `foo`. During the early device code splitting run
+of `tu_foo`, we may find that more than one of the resultant device images
+contain a definition for `foo`.
+
+We assert that any function definition for `foo` that is deemed a match by the
+ThinLTO infrastructure during the processing of `tu_bar` is valid.
+
+As a result of running early device code split, the fat object file generated as
+part of device compilation may contain multiple device code images.
+
+## Device code link time changes
+
+Before we go into the link time changes for SYCL, let's understand the device
+linking flow for AMD/NVIDIA devices:
+
+![Community linking flow](images/ThinLTOCommunityFlow.svg)
+
+SYCL has two differentiating requirements:
+
+1) The SPIR-V backend is not production ready and the SPIR-V translator is used.
+2) The SYCL runtime requires metadata (module properties and module symbol
+table) computed from device images that will be stored along the device images
+in the fat executable.
+
+The effect of requirement 1) is that instead of letting ThinLTO call the SPIR-V
+backend, we add a callback that runs right before CodeGen would run. In that
+callback, we call the SPIR-V translator and store the resultant file path for
+use later, and we instruct the ThinLTO framework to not perform CodeGen.
+
+An interesting additional fact about requirement 2) is that we actually need to
+process fully linked module to accurate compute the module properties. One
+example where we need the full module is to [compute the required devicelib
+mask](https://github.com/intel/llvm/blob/sycl/llvm/lib/SYCLLowerIR/SYCLDeviceLibReqMask.cpp).
+If we only process the device code that was included in the original fat object
+input to `clang-linker-wrapper`, we will miss devicelib calls in referenced
+`SYCL_EXTERNAL` functions.
+
+The effect of requirement 2) is that we store the fully linked device image for
+metadata computation in the SYCL-specific handing code after the ThinLTO
+framework has completed. Another option would be to try to compute the metadata
+inside the ThinLTO framework callbacks, but this would require SYCL-specific
+arguments to many caller functions in the stack and pollute community code.
+
+Here is the current ThinLTO flow for SYCL:
+
+![SYCL linking flow](images/ThinLTOSYCLFlow.svg)
+
+We add a `PreCodeGenModuleHook` function to the `LTOConfig` object so that we
+can process the fully linked module without running the backend.
+
+However, the flow is not ideal for many reasons:
+
+1) We are relying on the external `llvm-spirv` tool instead of the SPIR-V
+backend. We could slightly improve this issue by using a library call to the
+SPIR-V translator instead of the tool, however the library API requires setting
+up an object to represent the arguments while we only have strings, and it's
+non-trivial to parse the strings to figure out how to create the argument
```

Comment:
Am I right that you are talking about extensions here?

I don't think that it is necessarily non-trivial. If what we get from clang driver is what we should pass to `llvm-spirv`, then we can simply copy an existing function from `llvm-spirv` which does this parsing:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/tools/llvm-spirv/llvm-spirv.cpp#L516

Reply:
it's not just that, if it were the extensions it would be not that bad.

let's consider the flag `llvm-spirv` flag `spirv-max-version`. This is a `cl::opt` global variable and is set in `llvm-spirv`'s main by parsing argv as is standard.

In `clang-linker-wrapper`, this is just 
If we want to set the same flag, we would need to somehow get that same global variable set, either by externing it and seting it ourselves (which means we need to parse the strings from the driver and map them to a global variable as well as all the enum mappings for the options). Doing that for all options is what I meant by non-trivial. 

If I missed some obvious easy way of doing it let me know

Reply:
> let's consider the flag `llvm-spirv` flag `--spirv-max-version`. This is a `cl::opt` global variable `MaxSPIRVVersion` and is set in `llvm-spirv`'s main by parsing argv as is standard.

Do we have a public `clang` driver flag which allows to set maximum SPIR-V version? If not, do we pass different values of that option to the translator based on some other flags?

I guess my high-level point is: do we really have to pass all those flags through the linker-wrapper, or can we just come up with their values _within_ clang-linker wrapper based on the module itself?
That point of view stems from the fact that I don't really know which flags to the translator are currently "dynamic" and my assumption (which could be wrong one) that we don't have that many "dynamic" options which would actually require any handling.

Reply:
Sorry, ` --spirv-max-version` was just a random option I picked, that one in particular might always have the same value. But even if it does, we need to implement some way to get that global variable inside the `llvm-spirv` library set.
Even if `clang-linker-wrapper` knows all the options that need to be passed, we need to come up with how to set all the globals which seems flaky.

The consensuses from other reviewers here is to just use the SPIR-V backend now since it is apparently pretty good, so I'm not planning to use the translator anymore.

Reply:
You seem to be talking about how to initialize globals within `llvm-spirv/main.cpp` (or whatever  name of the file is), but that shouldn't be necessary. API entry points accept all those options via `TranslatorOpts` object. The translator library does not use any globals to access options

Reply:
Sorry, you're right, I'm too tried. Still, having to parse the strings from the driver and map it to the getter/setter for the object seems not ideal.